File: frontend/src/services/shareClassService.ts
import { api } from './api';
import type { 
  ShareClass, 
  CreateShareClassRequest, 
  UpdateShareClassRequest, 
  ShareClassStatistics 
} from '../types/company';

export class ShareClassService {
  /**
   * Get all active share classes for a company
   */
  static async getActiveShareClassesByCompany(companyId: number): Promise<ShareClass[]> {
    try {
      const response = await api.get(`/v1/companies/${companyId}/share-classes?activeOnly=true`);
      return response.data;
    } catch (error) {
      console.error('Error fetching active share classes:', error);
      throw error;
    }
  }

  /**
   * Get all share classes for a company (including inactive)
   */
  static async getAllShareClassesByCompany(companyId: number): Promise<ShareClass[]> {
    try {
      const response = await api.get(`/v1/companies/${companyId}/share-classes?activeOnly=false`);
      return response.data;
    } catch (error) {
      console.error('Error fetching all share classes:', error);
      throw error;
    }
  }

  /**
   * Get a specific share class by ID
   */
  static async getShareClassById(companyId: number, shareClassId: number): Promise<ShareClass> {
    try {
      const response = await api.get(`/v1/companies/${companyId}/share-classes/${shareClassId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching share class:', error);
      throw error;
    }
  }

  /**
   * Get share class by company ID and class code
   */
  static async getShareClassByCode(companyId: number, classCode: string): Promise<ShareClass> {
    try {
      const response = await api.get(`/v1/companies/${companyId}/share-classes/by-code/${classCode}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching share class by code:', error);
      throw error;
    }
  }

  /**
   * Create a new share class
   */
  static async createShareClass(companyId: number, request: CreateShareClassRequest): Promise<ShareClass> {
    try {
      const response = await api.post(`/v1/companies/${companyId}/share-classes`, request);
      return response.data;
    } catch (error) {
      console.error('Error creating share class:', error);
      throw error;
    }
  }

  /**
   * Update an existing share class
   */
  static async updateShareClass(
    companyId: number, 
    shareClassId: number, 
    request: UpdateShareClassRequest
  ): Promise<ShareClass> {
    try {
      const response = await api.put(`/v1/companies/${companyId}/share-classes/${shareClassId}`, request);
      return response.data;
    } catch (error) {
      console.error('Error updating share class:', error);
      throw error;
    }
  }

  /**
   * Deactivate a share class (soft delete)
   */
  static async deactivateShareClass(companyId: number, shareClassId: number): Promise<ShareClass> {
    try {
      const response = await api.delete(`/v1/companies/${companyId}/share-classes/${shareClassId}`);
      return response.data;
    } catch (error) {
      console.error('Error deactivating share class:', error);
      throw error;
    }
  }

  /**
   * Get share class statistics for a company
   */
  static async getShareClassStatistics(companyId: number): Promise<ShareClassStatistics[]> {
    try {
      const response = await api.get(`/v1/companies/${companyId}/share-classes/statistics`);
      return response.data;
    } catch (error) {
      console.error('Error fetching share class statistics:', error);
      throw error;
    }
  }

  /**
   * Validate share class data before creation
   */
  static validateCreateRequest(request: CreateShareClassRequest): string[] {
    const errors: string[] = [];

    if (!request.className || request.className.trim().length === 0) {
      errors.push('Class name is required');
    }
    if (request.className && request.className.length > 100) {
      errors.push('Class name cannot exceed 100 characters');
    }

    if (!request.classCode || request.classCode.trim().length === 0) {
      errors.push('Class code is required');
    }
    if (request.classCode && request.classCode.length > 20) {
      errors.push('Class code cannot exceed 20 characters');
    }

    if (request.description && request.description.length > 500) {
      errors.push('Description cannot exceed 500 characters');
    }

    // Voting rights validation
    if (request.votingRights && !['NONE', 'ORDINARY', 'WEIGHTED', 'RESTRICTED'].includes(request.votingRights)) {
      errors.push('Invalid voting rights value');
    }
    if (request.votingRights === 'NONE' && request.votesPerShare && request.votesPerShare > 0) {
      errors.push('Cannot have votes per share when voting rights is NONE');
    }
    if (request.votingRights && request.votingRights !== 'NONE' && (!request.votesPerShare || request.votesPerShare <= 0)) {
      errors.push('Must have positive votes per share when voting rights is not NONE');
    }

    // Dividend rights validation
    if (request.dividendRights && !['NONE', 'ORDINARY', 'PREFERRED', 'CUMULATIVE'].includes(request.dividendRights)) {
      errors.push('Invalid dividend rights value');
    }
    if (request.dividendRights && ['PREFERRED', 'CUMULATIVE'].includes(request.dividendRights) && !request.dividendRate) {
      errors.push('Dividend rate is required for preferred or cumulative dividend rights');
    }
    if (request.dividendRate && (request.dividendRate < 0 || request.dividendRate > 1)) {
      errors.push('Dividend rate must be between 0 and 1 (0% to 100%)');
    }

    // Capital distribution validation
    if (request.capitalDistributionRights && !['ORDINARY', 'PREFERRED', 'NONE'].includes(request.capitalDistributionRights)) {
      errors.push('Invalid capital distribution rights value');
    }

    // Par value validation
    if (request.isNoParValue && request.parValue !== undefined && request.parValue !== null) {
      errors.push('Cannot have par value when shares are designated as no par value');
    }
    if (request.parValue && request.parValue <= 0) {
      errors.push('Par value must be positive');
    }

    // Liquidation preference validation
    if (request.liquidationPreferenceMultiple && request.liquidationPreferenceMultiple <= 0) {
      errors.push('Liquidation preference multiple must be positive');
    }

    // Currency validation
    if (request.currency && request.currency.length !== 3) {
      errors.push('Currency code must be exactly 3 characters');
    }

    return errors;
  }

  /**
   * Format voting rights for display
   */
  static formatVotingRights(shareClass: ShareClass): string {
    switch (shareClass.votingRights) {
      case 'NONE':
        return 'No voting rights';
      case 'ORDINARY':
        return `${shareClass.votesPerShare} vote${shareClass.votesPerShare !== 1 ? 's' : ''} per share`;
      case 'WEIGHTED':
        return `${shareClass.votesPerShare} weighted vote${shareClass.votesPerShare !== 1 ? 's' : ''} per share`;
      case 'RESTRICTED':
        return `${shareClass.votesPerShare} restricted vote${shareClass.votesPerShare !== 1 ? 's' : ''} per share`;
      default:
        return 'Unknown';
    }
  }

  /**
   * Format dividend rights for display
   */
  static formatDividendRights(shareClass: ShareClass): string {
    switch (shareClass.dividendRights) {
      case 'NONE':
        return 'No dividend rights';
      case 'ORDINARY':
        return 'Ordinary dividend rights';
      case 'PREFERRED':
        const rate = shareClass.dividendRate ? (shareClass.dividendRate * 100).toFixed(2) : '0';
        return `${rate}% preferred dividend`;
      case 'CUMULATIVE':
        const cumulativeRate = shareClass.dividendRate ? (shareClass.dividendRate * 100).toFixed(2) : '0';
        return `${cumulativeRate}% cumulative dividend`;
      default:
        return 'Unknown';
    }
  }

  /**
   * Format liquidation preference for display
   */
  static formatLiquidationPreference(shareClass: ShareClass): string {
    if (shareClass.capitalDistributionRights === 'PREFERRED' && shareClass.liquidationPreferenceMultiple) {
      return `${shareClass.liquidationPreferenceMultiple}x liquidation preference`;
    }
    return shareClass.capitalDistributionRights === 'ORDINARY' ? 'Pro rata participation' : 'No liquidation preference';
  }

  /**
   * Get key features summary for a share class
   */
  static getKeyFeatures(shareClass: ShareClass): string[] {
    const features: string[] = [];

    if (shareClass.hasVotingRights) {
      features.push(this.formatVotingRights(shareClass));
    }

    if (shareClass.hasDividendRights) {
      features.push(this.formatDividendRights(shareClass));
    }

    if (shareClass.hasLiquidationPreference) {
      features.push(this.formatLiquidationPreference(shareClass));
    }

    if (shareClass.isRedeemable) {
      features.push('Redeemable');
    }

    if (shareClass.isConvertible) {
      features.push('Convertible');
    }

    if (shareClass.requiresBoardApproval) {
      features.push('Board approval required for transfers');
    }

    if (shareClass.hasPreemptiveRights) {
      features.push('Preemptive rights');
    }

    if (shareClass.hasTagAlongRights) {
      features.push('Tag-along rights');
    }

    if (shareClass.hasDragAlongRights) {
      features.push('Drag-along rights');
    }

    return features.length > 0 ? features : ['Standard ordinary shares'];
  }
}

File: backend/src/main/kotlin/nz/govt/companiesoffice/register/service/ShareClassService.kt
package nz.govt.companiesoffice.register.service

import nz.govt.companiesoffice.register.audit.AuditAction
import nz.govt.companiesoffice.register.audit.AuditService
import nz.govt.companiesoffice.register.entity.Company
import nz.govt.companiesoffice.register.entity.ShareClass
import nz.govt.companiesoffice.register.repository.CompanyRepository
import nz.govt.companiesoffice.register.repository.ShareClassRepository
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.math.BigDecimal
import java.time.LocalDateTime

@Service
@Transactional
class ShareClassService(
    private val shareClassRepository: ShareClassRepository,
    private val companyRepository: CompanyRepository,
    private val auditService: AuditService,
) {
    private val logger = LoggerFactory.getLogger(ShareClassService::class.java)

    /**
     * Create a new share class for a company
     */
    fun createShareClass(
        companyId: Long,
        className: String,
        classCode: String,
        description: String? = null,
        votingRights: String = "NONE",
        votesPerShare: Int = 0,
        dividendRights: String = "NONE",
        dividendRate: BigDecimal? = null,
        isCumulativeDividend: Boolean = false,
        dividendPriority: Int = 0,
        capitalDistributionRights: String = "ORDINARY",
        liquidationPreferenceMultiple: BigDecimal? = null,
        liquidationPriority: Int = 0,
        isRedeemable: Boolean = false,
        isConvertible: Boolean = false,
        parValue: BigDecimal? = null,
        isNoParValue: Boolean = false,
        currency: String = "NZD",
        isTransferable: Boolean = true,
        transferRestrictions: String? = null,
        requiresBoardApproval: Boolean = false,
        hasPreemptiveRights: Boolean = false,
        hasTagAlongRights: Boolean = false,
        hasDragAlongRights: Boolean = false,
    ): ShareClass {
        logger.info("Creating share class for company {}: {} ({})", companyId, className, classCode)

        val company = companyRepository.findById(companyId).orElseThrow {
            IllegalArgumentException("Company not found with ID: $companyId")
        }

        // Validate uniqueness
        if (shareClassRepository.existsByCompany_IdAndClassCode(companyId, classCode)) {
            throw IllegalArgumentException("Share class code '$classCode' already exists for this company")
        }

        if (shareClassRepository.existsByCompany_IdAndClassName(companyId, className)) {
            throw IllegalArgumentException("Share class name '$className' already exists for this company")
        }

        // Validate business rules
        validateShareClassRules(
            votingRights,
            votesPerShare,
            dividendRights,
            dividendRate,
            isCumulativeDividend,
            parValue,
            isNoParValue,
            liquidationPreferenceMultiple,
        )

        val shareClass = ShareClass(
            company = company,
            className = className.trim(),
            classCode = classCode.trim().uppercase(),
            description = description?.trim(),
            votingRights = votingRights,
            votesPerShare = votesPerShare,
            dividendRights = dividendRights,
            dividendRate = dividendRate,
            isCumulativeDividend = isCumulativeDividend,
            dividendPriority = dividendPriority,
            capitalDistributionRights = capitalDistributionRights,
            liquidationPreferenceMultiple = liquidationPreferenceMultiple,
            liquidationPriority = liquidationPriority,
            isRedeemable = isRedeemable,
            isConvertible = isConvertible,
            parValue = parValue,
            isNoParValue = isNoParValue,
            currency = currency,
            isTransferable = isTransferable,
            transferRestrictions = transferRestrictions?.trim(),
            requiresBoardApproval = requiresBoardApproval,
            hasPreemptiveRights = hasPreemptiveRights,
            hasTagAlongRights = hasTagAlongRights,
            hasDragAlongRights = hasDragAlongRights,
        )

        val savedShareClass = shareClassRepository.save(shareClass)

        auditService.logEvent(
            action = AuditAction.CREATE,
            resourceType = "ShareClass",
            resourceId = savedShareClass.id.toString(),
            details = mapOf(
                "className" to className,
                "classCode" to classCode,
                "companyId" to company.id,
                "companyName" to company.companyName,
            ),
        )

        logger.info("Successfully created share class {} for company {}", savedShareClass.id, companyId)
        return savedShareClass
    }

    /**
     * Update an existing share class
     */
    fun updateShareClass(
        shareClassId: Long,
        className: String? = null,
        description: String? = null,
        votingRights: String? = null,
        votesPerShare: Int? = null,
        dividendRights: String? = null,
        dividendRate: BigDecimal? = null,
        isCumulativeDividend: Boolean? = null,
        dividendPriority: Int? = null,
        capitalDistributionRights: String? = null,
        liquidationPreferenceMultiple: BigDecimal? = null,
        liquidationPriority: Int? = null,
        transferRestrictions: String? = null,
        requiresBoardApproval: Boolean? = null,
        hasPreemptiveRights: Boolean? = null,
        hasTagAlongRights: Boolean? = null,
        hasDragAlongRights: Boolean? = null,
    ): ShareClass {
        logger.info("Updating share class {}", shareClassId)

        val shareClass = shareClassRepository.findById(shareClassId).orElseThrow {
            IllegalArgumentException("Share class not found with ID: $shareClassId")
        }

        // Check if class name uniqueness would be violated
        if (className != null && className != shareClass.className) {
            if (shareClassRepository.existsByCompany_IdAndClassName(shareClass.company.id, className)) {
                throw IllegalArgumentException("Share class name '$className' already exists for this company")
            }
            shareClass.className = className.trim()
        }

        // Update fields if provided
        description?.let { shareClass.description = it.trim() }
        votingRights?.let { shareClass.votingRights = it }
        votesPerShare?.let { shareClass.votesPerShare = it }
        dividendRights?.let { shareClass.dividendRights = it }
        dividendRate?.let { shareClass.dividendRate = it }
        isCumulativeDividend?.let { shareClass.isCumulativeDividend = it }
        dividendPriority?.let { shareClass.dividendPriority = it }
        capitalDistributionRights?.let { shareClass.capitalDistributionRights = it }
        liquidationPreferenceMultiple?.let { shareClass.liquidationPreferenceMultiple = it }
        liquidationPriority?.let { shareClass.liquidationPriority = it }
        transferRestrictions?.let { shareClass.transferRestrictions = it.trim() }
        requiresBoardApproval?.let { shareClass.requiresBoardApproval = it }
        hasPreemptiveRights?.let { shareClass.hasPreemptiveRights = it }
        hasTagAlongRights?.let { shareClass.hasTagAlongRights = it }
        hasDragAlongRights?.let { shareClass.hasDragAlongRights = it }

        shareClass.updatedAt = LocalDateTime.now()

        // Validate business rules
        validateShareClassRules(
            shareClass.votingRights,
            shareClass.votesPerShare,
            shareClass.dividendRights,
            shareClass.dividendRate,
            shareClass.isCumulativeDividend,
            shareClass.parValue,
            shareClass.isNoParValue,
            shareClass.liquidationPreferenceMultiple,
        )

        val savedShareClass = shareClassRepository.save(shareClass)

        auditService.logEvent(
            action = AuditAction.UPDATE,
            resourceType = "ShareClass",
            resourceId = savedShareClass.id.toString(),
            details = mapOf(
                "className" to savedShareClass.className,
                "classCode" to savedShareClass.classCode,
                "companyId" to savedShareClass.company.id,
            ),
        )

        logger.info("Successfully updated share class {}", shareClassId)
        return savedShareClass
    }

    /**
     * Get share class by ID
     */
    @Transactional(readOnly = true)
    fun getShareClassById(shareClassId: Long): ShareClass? {
        return shareClassRepository.findById(shareClassId).orElse(null)
    }

    /**
     * Get all active share classes for a company
     */
    @Transactional(readOnly = true)
    fun getActiveShareClassesByCompany(companyId: Long): List<ShareClass> {
        return shareClassRepository.findByCompany_IdAndIsActiveTrue(companyId)
    }

    /**
     * Get all share classes for a company (including inactive)
     */
    @Transactional(readOnly = true)
    fun getAllShareClassesByCompany(companyId: Long): List<ShareClass> {
        return shareClassRepository.findByCompany_IdOrderByCreatedAtAsc(companyId)
    }

    /**
     * Get share class by company and class code
     */
    @Transactional(readOnly = true)
    fun getShareClassByCode(companyId: Long, classCode: String): ShareClass? {
        return shareClassRepository.findByCompany_IdAndClassCode(companyId, classCode)
    }

    /**
     * Deactivate a share class (soft delete)
     */
    fun deactivateShareClass(shareClassId: Long): ShareClass {
        logger.info("Deactivating share class {}", shareClassId)

        val shareClass = shareClassRepository.findById(shareClassId).orElseThrow {
            IllegalArgumentException("Share class not found with ID: $shareClassId")
        }

        // Check if there are active share allocations for this class
        // This would need to be implemented in ShareAllocationService
        // For now, we'll allow deactivation but log a warning

        shareClass.isActive = false
        shareClass.updatedAt = LocalDateTime.now()

        val savedShareClass = shareClassRepository.save(shareClass)

        auditService.logEvent(
            action = AuditAction.DELETE,
            resourceType = "ShareClass",
            resourceId = savedShareClass.id.toString(),
            details = mapOf(
                "className" to savedShareClass.className,
                "classCode" to savedShareClass.classCode,
                "companyId" to savedShareClass.company.id,
                "action" to "deactivation",
            ),
        )

        logger.info("Successfully deactivated share class {}", shareClassId)
        return savedShareClass
    }

    /**
     * Get share class statistics for a company
     */
    @Transactional(readOnly = true)
    fun getShareClassStatistics(companyId: Long): List<Map<String, Any>> {
        val stats = shareClassRepository.getShareClassStatistics(companyId)
        return stats.map { row ->
            mapOf(
                "shareClassId" to row[0],
                "className" to row[1],
                "classCode" to row[2],
                "allocationCount" to row[3],
                "totalShares" to row[4],
                "totalValue" to row[5],
            )
        }
    }

    /**
     * Create default ordinary share class for a company
     */
    fun createDefaultOrdinaryShares(company: Company): ShareClass {
        logger.info("Creating default ordinary share class for company {}", company.id)

        return createShareClass(
            companyId = company.id,
            className = "Ordinary Shares",
            classCode = "ORDINARY",
            description = "Standard ordinary shares with full voting rights and participation in dividends",
            votingRights = "ORDINARY",
            votesPerShare = 1,
            dividendRights = "ORDINARY",
            capitalDistributionRights = "ORDINARY",
            isTransferable = true,
            hasPreemptiveRights = true,
        )
    }

    /**
     * Validate share class business rules
     */
    private fun validateShareClassRules(
        votingRights: String,
        votesPerShare: Int,
        dividendRights: String,
        dividendRate: BigDecimal?,
        isCumulativeDividend: Boolean,
        parValue: BigDecimal?,
        isNoParValue: Boolean,
        liquidationPreferenceMultiple: BigDecimal?,
    ) {
        // Voting rights consistency
        if (votingRights == "NONE" && votesPerShare > 0) {
            throw IllegalArgumentException("Cannot have votes per share when voting rights is NONE")
        }
        if (votingRights != "NONE" && votesPerShare <= 0) {
            throw IllegalArgumentException("Must have positive votes per share when voting rights is not NONE")
        }

        // Dividend rate consistency
        if (dividendRights in listOf("PREFERRED", "CUMULATIVE") && dividendRate == null) {
            throw IllegalArgumentException("Dividend rate is required for preferred or cumulative dividend rights")
        }
        if (dividendRate != null && (dividendRate < BigDecimal.ZERO || dividendRate > BigDecimal.ONE)) {
            throw IllegalArgumentException("Dividend rate must be between 0 and 1 (0% to 100%)")
        }

        // Par value consistency
        if (isNoParValue && parValue != null) {
            throw IllegalArgumentException("Cannot have par value when shares are designated as no par value")
        }
        if (parValue != null && parValue <= BigDecimal.ZERO) {
            throw IllegalArgumentException("Par value must be positive")
        }

        // Liquidation preference consistency
        if (liquidationPreferenceMultiple != null && liquidationPreferenceMultiple <= BigDecimal.ZERO) {
            throw IllegalArgumentException("Liquidation preference multiple must be positive")
        }
    }
}
